src/Processors/Transforms/MemoryBoundMerging.h

```diff
@@ -0,0 +1,213 @@
+        {
+            for (auto & input : inputs)
+                input.close();
+
+            return Status::Finished;
+        }
+
+        if (!output.canPush())
+        {
+            for (auto & input : inputs)
+                input.setNotNeeded();
+
+            return Status::PortFull;
+        }
+
+        /// Push if have chunk that is the next in order
+        bool pushed_to_output = tryPushChunk();
+
+        bool need_data = false;
+        bool all_finished = true;
+
+        /// Try read new chunk
+        auto in = inputs.begin();
+        for (size_t input_num = 0; input_num < num_inputs; ++input_num, ++in)
+        {
+            if (in->isFinished())
+            {
+                is_input_finished[input_num] = true;
+                continue;
+            }
+
+            /// We want to keep not more than `num_inputs` chunks in memory (and there will be only a single chunk with the given (bucket_id, chunk_num)).
+            const bool bucket_from_this_input_still_in_memory = chunks.contains(last_chunk_id[input_num]);
+            if (bucket_from_this_input_still_in_memory)
+            {
+                all_finished = false;
+                continue;
+            }
+
+            in->setNeeded();
+
+            if (!in->hasData())
+            {
+                need_data = true;
+                all_finished = false;
+                continue;
+            }
+
+            auto chunk = in->pull();
+            addChunk(std::move(chunk), input_num);
+
+            if (in->isFinished())
+            {
+                is_input_finished[input_num] = true;
+            }
+            else
+            {
+                /// If chunk was pulled, then we need data from this port.
+                need_data = true;
+                all_finished = false;
+            }
+        }
+
+        if (pushed_to_output)
+            return Status::PortFull;
+
+        if (tryPushChunk())
+            return Status::PortFull;
+
+        if (need_data)
+            return Status::NeedData;
+
+        if (!all_finished)
+            throw Exception(
+                "SortingAggregatedForMemoryBoundMergingTransform has read bucket, but couldn't push it.", ErrorCodes::LOGICAL_ERROR);
+
+        if (overflow_chunk)
+        {
+            output.push(std::move(overflow_chunk));
+            return Status::PortFull;
+        }
+
+        output.finish();
+        return Status::Finished;
+    }
+
+private:
+    bool tryPushChunk()
+    {
+        auto & output = outputs.front();
+
+        if (chunks.empty())
+            return false;
+
+        /// Chunk with min id
+        auto it = chunks.begin();
+        auto current_chunk_id = it->first;
+
+        /// Check if it is actually next in order
+        for (size_t input = 0; input < num_inputs; ++input)
+            if (!is_input_finished[input] && last_chunk_id[input] < current_chunk_id)
+                return false;
+
+        LOG_DEBUG(
+            &Poco::Logger::get("debug"),
```

Comment:
LOG_TRACE at least.
Logger name is wrong, let's replace to `&Poco::Logger::get("SortingAggregatedForMemoryBoundMergingTransform"), "Pushing chunk with id"`.
It's better to save logger in a class member.

Reply:
it is too noisy imo for trace logs, so I just remove it, 'cause it was needed mostly for debug during implementation